File: Assign4/cpp/problem02.cpp
#include <ctime>
#include <iostream>
#include <limits>
#include <vector>

using namespace std;

int dp[100];

void generate_matrix(int row, int column, vector<vector<int>> &my_matrix) {
  for (int i = 0; i < row; i++) {
    my_matrix.push_back(vector<int>(column, 0));
    for (int j = 0; j < column; j++) {
      my_matrix[i][j] = rand() % 100 + 1;
    }
  }
}

void print_matrix(vector<vector<int>> &input_matrix) {
  int row = input_matrix.size();
  int column = input_matrix[0].size();
  cout << "output of matrix " << row << " by " << column << "\n";
  for (int i = 0; i < row; i++) {
    for (int j = 0; j < column; j++) {
      cout << input_matrix[i][j] << " ";
    }
    cout << "\n";
  }
  cout << "\n";
}

void matrix_chain_order(vector<int> dim) {
  int n = dim.size() - 1;
  vector<vector<float>> m(n + 1, vector<float>(n + 1, 0));
  vector<vector<float>> s(n + 1, vector<float>(n + 1, 0));

  for (int l = 2; l < n + 1; l++) {
    for (int i = 1; i < n - l + 1; i++) {
      int j = i + l - 1;
      m[i][j] = numeric_limits<float>::infinity();
      for (int k = i; k < j; k++) {
        float q = m[i][k] + m[k + 1][j] + dim[i - 1] * dim[k] * dim[j];
        if (q < m[i][j]) {
          m[i][j] = q;
          s[i][j] = k;
        }
      }
    }
  }
}

vector<vector<int>> matrix_multi(vector<vector<int>> &f_matrix,
                                 vector<vector<int>> &s_matrix) {
  print_matrix(f_matrix);
  print_matrix(s_matrix);
}

int main(void) {
  ios::sync_with_stdio(false);
  cin.tie(NULL);
  cout.tie(NULL);
  srand(time(NULL));
  vector<vector<int>> matrix_5x3;
  vector<vector<int>> matrix_3x7;
  vector<vector<int>> matrix_7x10;
  generate_matrix(5, 3, matrix_5x3);
  generate_matrix(3, 7, matrix_3x7);
  generate_matrix(7, 10, matrix_7x10);

  matrix_multi(matrix_5x3, matrix_3x7);

  return 0;
}

File: Assign4/python/problem02.py
import random


def generate_matrix(row, column):
    matrix = [[0 for i in range(column)] for j in range(row)]
    for i in range(row):
        for j in range(column):
            matrix[i][j] = random.randrange(1, 100)
    return matrix


def print_matrix(input_matrix):
    row = len(input_matrix)
    column = len(input_matrix[0])
    print("output of matrix", row, " by ", column)
    for i in range(row):
        for j in range(column):
            print(input_matrix[i][j], end=" ")
        print('')
    print('')


def matrix_chain_order(p):
    n = len(p) - 1
    m = [[0 for i in range(n+1)] for j in range(n+1)]
    s = [[0 for i in range(n+1)] for j in range(n+1)]

    for l in range(2, n+1):
        for i in range(1, n - l + 2):
            j = i + l - 1
            m[i][j] = float("inf")
            for k in range(i, j):
                q = m[i][k] + m[k + 1][j] + p[i-1] * p[k] * p[j]
                if q < m[i][j]:
                    m[i][j] = q
                    s[i][j] = k

    return (m, s)


def opt_chain_order(s, i, j):
    if i == j:
        print("A", end="")
        return
    k = s[i][j]
    print("(", end="")
    opt_chain_order(s, i, k)
    opt_chain_order(s, k+1, j)
    print(")", end="")


def matrix_multi(first, second):
    if len(first[0]) != len(second):
        print("incompatible dimensions")
        return
    result = [[0 for i in range(len(second[0]))] for j in range(len(first))]
    for i in range(0, len(first)):
        for j in range(0, len(second[0])):
            for k in range(0, len(first[0])):
                result[i][j] += first[i][k] * second[k][j]
    return result


p = [5, 3, 7, 10]
n = len(p) - 1
matrix_5x3 = generate_matrix(5, 3)
matrix_3x7 = generate_matrix(3, 7)
matrix_7x10 = generate_matrix(7, 10)

m, s = matrix_chain_order(p)
print(m)
print(s)

print("optimal chain order is ", end=" ")
opt_chain_order(s, 1, n)
print('')
print('')
print("minimal number of computation is ", m[1][n])
print('')
c = matrix_multi(matrix_3x7, matrix_7x10)
d = matrix_multi(matrix_5x3, c)
print_matrix(d)
